### PR TITLE
feat(runtime): enhance the access protection security model

### LINKED ISSUE #448: Enhance the access protection security model
### Challenge
The current access protection model knows two types of access: `public` and `private`. When setting the deployment model to a scaled model, these types are not enough.

`Public` functions can always be called from any segment, regardless of their location (front end or backend). Private functions cannot be called from any other segment. This model does not support the use case where trusted clients, i.e. backend nodes, can call functions from other segments, but untrusted clients, i.e. front ends, cannot.

To support this use case we need to extend the access protection model. A `protected` identifier could be added to fuctions to indicate that these can be called from trusted clients, but not from any untrusted client.

### Solution description

A mechanism to trust clients need to be implemented. We could use a simple strategy where a sectet / trust key is configured in the gateway / node configuration, and checked during startup. The configuration could look like the following.

```js
// services/gateway.json
{
    "gateway":
    {
        "SECRET_KEY": "ThisVerySecretKey"
    }
}
```

For each trusted client, we also need to set this key in the node configuration.

```js
// services/node.json
{
    "node":
    {
        "TRUST_KEY": "ThisVerySecretKey"
    }
}
```

When the gateway starts, the key is stored in the configuration and used to verify incoming requests on the `/nodes` path.

When a node tries to connect to the gateway on startup, the gateway needs to check if the key that is provided is valid. If the secret is not known, then the node won't be allowed and a forbidden will be returned. In case the node doesn't have a key, it will be allowed as an untrusted client.

After a node is started, the gateway could provide a new secret to the node for access to protected functions, or we can start with the same key for now. The front end is not able to register itself as a node, and thus won't be able to be acknowledged as a trusted client. And any regular call to the gateway won't reveal the secret.

The trust of a client is not kept as a list in the gateway for now. The gateway will look for the header `X-Access-Key` and checks this key with its secret key. If that is a match, the `protected` functions will also be available for execution. If the header is missing, then only the `public` fuctions are considered. And in case the key is incorrect, the gateway will return a forbidden exception.

To get the key to the gateway, the `Remote` needs to set this header. It will set the `X-Access-Key` header with the value of the trust key from the configuration, or the new secret that the gateway returned at startup. A fixed key has the advantage that if the gateway gets restarted for whatever reason, the keys are still valid and not all nodes need to get restarted as well.

To indicate a function is `protected`, the keyword needs to be added to the list of acceptable values in the segment configuration.
```js
// segments/node.json
{
    "./domain":
    {
        "sayPublic": {  "access": "public" },
        "sayPrivate": {  "access": "private" },
        "sayProtected": {  "access": "protected" }
    }
}
```

### PR DESCRIPTION
**Pull request by BasMasking**
Fixes #448

Changes proposed in this pull request:
- added option to created trusted clients
- added protected access for procedures
- updated documentation
- updated example to reflect the change

@MaskingTechnology/jitar

### CODE REVIEW

File: examples/concepts/access-protection/segments/game.segment.json
Comment (petermasking) [inline examples/concepts/access-protection/segments/game.segment.json:1]: Is it more clear to rename this segment to `protected.segment.json`?

File: examples/concepts/access-protection/segments/web.segment.json
Comment (petermasking) [inline examples/concepts/access-protection/segments/web.segment.json:1]: Is it more clear to rename this segment to `public.segment.json`

File: packages/runtime/src/models/Procedure.ts [hunk 1]
```diff
@@ -25,6 +25,22 @@ export default class Procedure
         return implementations.some(implementation => implementation.public);
     }
 
+    get protected()
+    {
```
Comment (petermasking) [inline packages/runtime/src/models/Procedure.ts:29]: Should we mention this in the docs?

File: packages/runtime/src/services/LocalNode.ts [hunk 2]
```diff
@@ -130,12 +136,30 @@ export default class LocalNode extends Node
...
+            const headers = request.headers;
+            headers.set('x-access-key', this.#secret);
+        }
 
         return this.#gateway.run(request);
     }
 
     async #runProcedure(procedure: Procedure, request: Request): Promise<Response>
     {
+        const secret = request.headers.get('x-access-key');
```
Comment (petermasking) [inline packages/runtime/src/services/LocalNode.ts:151]:
```suggestion
        const secret = request.getHeader('x-access-key');
```

File: packages/runtime/test/services/LocalGateway.spec.ts
Comment (petermasking) [inline packages/runtime/test/services/LocalGateway.spec.ts:1]: I'm missing a test for a failing registration of a node, and protected related test in the local node spec.

**Review by petermasking — CHANGES_REQUESTED** [review]

File: documentation/docs/deploy/segmentation.md [hunk 3]
```diff
@@ -80,6 +80,16 @@ The import names must correspond with the export names in the module. You can al
 
 Imports have multiple properties that can be configured. These properties will be explained next.
 
+### Trusted clients
+
+When building a distributed application, you don't want all functions to be available by the outside world. Some functions are only used internally by other segments. To protect the access to these functions, Jitar provides a `secret` property in the [runtime services](../fundamentals/runtime-services#node). This secret is used to create trusted clients. Trusted clients can access functions with the `protected` access level.
+
+Any client that wants to access a protected function must provide a valid access key. The access key needs to be added to the http header `x-access-key`. Any node that has a valid access key is considered a trusted client, and automatically adds the access key to the http header of outgoing requests. Any node that doesn't have a valid access key is considered an untrusted client and can only access `public` functions.
```
[the remark this answers is not in the document]
Comment (BasMasking) [reply documentation/docs/deploy/segmentation.md:87]: Good suggestion, I changed the header key.

File: documentation/docs/fundamentals/runtime-services.md [hunk 4]
```diff
@@ -115,6 +115,7 @@ The following configuration properties are available:
 * repository - url of the repository (required).
 * segments - list of segment names to load (optional, loads all segments by default).
 * middlewares - list of [middleware modules](../develop/middleware.md) to load (optional).
+* secret - shared secret for creating trusted clients (optional).
```
[the remark this answers is not in the document]
Comment (BasMasking) [reply documentation/docs/fundamentals/runtime-services.md:118]: It's a string to access the protected procedures, so I go for the `protectedAccessKey` for now. Any suggestions are welcome.

**Review by petermasking — CHANGES_REQUESTED** [review]
Besides the other review comments I'm still missing the tests for the local node. At least for running a protected procedure with a key, with an invalid key and a valid key.

File: examples/concepts/access-protection/requests.http [hunk 5]
```diff
@@ -1,6 +1,17 @@
 
 // Run the public function (succeeds)
 
+GET http://localhost:3000/rpc/web/guessSecret?secret=123
+
+###
+
+// Run the protected function with key (succeeds)
+GET http://localhost:3000/rpc/game/checkSecret?secret=123 HTTP/1.1
+X-Jitar-Protected-Access-Key: VERY_SECRET_KEY
```
Comment (petermasking) [inline examples/concepts/access-protection/requests.http:10]:
```suggestion
X-Jitar-Trust-Key: VERY_SECRET_KEY
```

File: documentation/docs/deploy/segmentation.md [hunk 6]
```diff
@@ -80,6 +80,16 @@ The import names must correspond with the export names in the module. You can al
 
 Imports have multiple properties that can be configured. These properties will be explained next.
 
+### Trusted clients
+
+When building a distributed application, you don't want all functions to be available by the outside world. Some functions are only used internally by other segments. To protect the access to these functions, Jitar provides a `trustKey` property in the [runtime services](../fundamentals/runtime-services#node). This key is used to create trusted clients. Trusted clients can access functions with the `protected` access level.
+
+Any client that wants to access a protected function must provide a valid key. It needs to be added to the http header `X-Jitar-Protected-Access-Key`. Any node that has a valid key is automatically considered a trusted client, and adds the access key to the http header of outgoing requests. Any node that doesn't have a valid access key is considered an untrusted client and can only access `public` functions.
```
[the remark this answers is not in the document]
Comment (petermasking) [reply documentation/docs/deploy/segmentation.md:87]: The key has been changed in the code, but not in the docs yet.

**Review by petermasking — CHANGES_REQUESTED** [review]

**Review by petermasking — APPROVED** [review]

**Pull request merged**